sql/src/main/java/io/crate/execution/engine/indexing/ShardDMLExecutor.java

```diff
@@ -155,12 +160,78 @@ private static String getLocalNodeId(ClusterService clusterService) {
         return nodeId;
     }
 
-    private static long processShardResponse(ShardResponse shardResponse) {
+    private static List<Object[]> intialResultRows() {
+        return new ArrayList<>();
+    }
+
+    private static long initialRowCount() {
+        return 0L;
+    }
+
+    private static long processRowCount(ShardResponse shardResponse) {
+        return processResponse(shardResponse, ShardResponse::successRowCount);
+    }
+
+    private static long combineRowCounts(long a, long b) {
+        return a + b;
+    }
+
+    private static Iterable<Row> finishRowCount(Long rowCount) {
+        return Collections.singletonList(new Row1(rowCount == null ? 0L : rowCount));
+    }
+
+    private static List<Object[]> combineResultRows(List<Object[]> a, List<Object[]> b) {
+        var result = new ArrayList<Object[]>();
+        result.addAll(a);
+        result.addAll(b);
+        return result;
+    }
+
+    private static Iterable<Row> finishResultRows(List<Object[]> result) {
+        return Lists2.map(result, RowN::new);
+    }
+
+    private static List<Object[]> processResulRows(ShardResponse shardResponse) {
+        return processResponse(shardResponse, ShardResponse::getResultRows);
+    }
+
+    private static <A> A processResponse(ShardResponse shardResponse, Function<ShardResponse, A> f) {
         Exception failure = shardResponse.failure();
         if (failure != null) {
             Throwables.throwIfUnchecked(failure);
             throw new RuntimeException(failure);
         }
-        return shardResponse.successRowCount();
+        return f.apply(shardResponse);
+    }
+
+    public static final ResultProcessing<Long> ROW_COUNT_PROCESSING = new ResultProcessing<>(
+        ShardDMLExecutor.initialRowCount(),
+        ShardDMLExecutor::processRowCount,
+        ShardDMLExecutor::combineRowCounts,
+        ShardDMLExecutor::finishRowCount
+    );
+
+    public static final ResultProcessing<List<Object[]>> RESULT_ROW_PROCESSING = new ResultProcessing<>(
+        ShardDMLExecutor.intialResultRows(),
+        ShardDMLExecutor::processResulRows,
+        ShardDMLExecutor::combineResultRows,
+        ShardDMLExecutor::finishResultRows
+    );
+
+    public static class ResultProcessing<Result> {
```

Comment:
I am not really happy with this solution, it smells like `MultiActionListener` but it is still better than moving 8 lambdas across.